Implement a simple LCD menu

numo68 opened the pull request:
# What does this implement/fix?

The integration implements a hierarchical menu on (for now) character-based LCD displays, primarily meant to be used together with a rotary encoder with a single push button. It is possible to navigate the menu, changing enum values (named integer) and float ones, with the values themselves being maintained in globals (that allows to use them freely in other places, persisting them etc). Automations are added to perform user-defined functionality on entering or leaving the menu level and changing the value or triggering a command. The menu can be hidden and shown again, enabling alternating between the menu and other usage for the display.

This is a first shot at the functionality and I am more searching for the feedback than aiming to merge right away.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2029

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml

# Example config.yaml
lcd_menu:
  id: my_lcd_menu
  display_id: my_lcd
  #mark_back: 0x08
  #mark_selected: 0x3e
  #mark_editing: 0x2a
  #mark_submenu: 0x7e
  active: True
  on_enter:
    then:
      lambda: 'ESP_LOGI("lcd_menu", "root enter");'
  on_leave:
    then:
      lambda: 'ESP_LOGI("lcd_menu", "root leave");'
  items:
    - type: back
      text: 'Back'
    - type: label
      text: 'Label 1'
    - type: menu
      text: 'Submenu 1'
      on_enter:
        then:
          lambda: 'ESP_LOGI("lcd_menu", "enter: %s", it->get_text().c_str());'
      on_leave:
        then:
          lambda: 'ESP_LOGI("lcd_menu", "leave: %s", it->get_text().c_str());'
      items:
        - type: back
          text: 'Back'
        - type: menu
          text: 'Submenu 21'
          items:
            - type: back
              text: 'Back'
            - type: command
              text: 'Show Main'
              on_value:
                then:
                  - display_menu.show_main:
        - type: label
          text: 'Label 22'
    - type: select
      text: 'Enum Item'
      immediate_edit: True
      select: my_enum_1
      on_enter:
        then:
          lambda: 'ESP_LOGI("lcd_menu", "select enter: %s, %s", it->get_text().c_str(), it->get_value_text().c_str());'
      on_leave:
        then:
          lambda: 'ESP_LOGI("lcd_menu", "select leave: %s, %s", it->get_text().c_str(), it->get_value_text().c_str());'
      on_value:
        then:
          lambda: 'ESP_LOGI("lcd_menu", "select value: %s, %s", it->get_text().c_str(), it->get_value_text().c_str());'
    - type: number
      text: 'Number'
      format: '%.2f'
      number: my_number_1
      on_value:
        then:
          lambda: 'ESP_LOGI("lcd_menu", "number value: %s, %s", it->get_text().c_str(), it->get_value_text().c_str());'
    - type: command
      text: 'Command'
      on_value:
        then:
          lambda: 'ESP_LOGI("lcd_menu", "command: %s", it->get_text().c_str());'  
    - type: switch
      text: 'Switch'
      switch: my_switch_1
      on_text: 'Bright'
      off_text: 'Dark'
      immediate_edit: False
      on_value:
        then:
          lambda: 'ESP_LOGI("lcd_menu", "switch value: %s", it->get_value_text().c_str());'  
    - type: command
      text: 'Hide'
      on_value:
        then:
          - display_menu.hide:  
    - type: custom
      text: !lambda 'return "Custom";'
      value_lambda: 'return "Val";'
      on_next:
        then:
          lambda: 'ESP_LOGI("lcd_menu", "custom next: %s", it->get_text().c_str());'
      on_prev:
        then:
          lambda: 'ESP_LOGI("lcd_menu", "custom prev: %s", it->get_text().c_str());'
    
number:
  - platform: template
    id: my_number_1
    optimistic: True
    min_value: 10.0
    max_value: 20.0
    step: 0.5

select:
  - platform: template
    id: my_enum_1
    optimistic: True
    options:
      - 'Red'
      - 'Green'
      - 'Blue'

switch:
  - platform: template
    id: my_switch_1
    optimistic: True

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

numo68 committed:
Implement a simple LCD menu

numo68 committed:
Fix formatting issues

numo68 committed:
Fix another lint issue

nagyrobi commented:
Whow. Awesome.

jesserockz commented:
I love the idea, But I think you can make it simpler by instead of having `enum` and `number` types, the user can reference a `select`, `number`, `switch` entities within yaml which would populate the valid options and have an easier to access place for storing the current value.

numo68 commented:
> But I think you can make it simpler by instead of having `enum` and `number` types, the user can reference a `select`, `number`, `switch` entities within yaml which would populate the valid options and have an easier to access place for storing the current value.

Thanks for your comments. Yes, my gut feeling regarding the globals is also that that does not feel completely right. The rationale for using it anyway especially for the `enum` was

1. when creating a menu to edit something I probably want to see the options in the menu item definition rather than at some opposite end of the file. Sure, this can be debated, another view is that this is basically a typed value and the type definition is at the `select`.
2. from the bird's eye view the `select` is an *unordered* set, there is no concept of cycling through options built in. One would need to find it in the list (by the way, `SelectTraits::get_options()` is not returning a reference) and work from there.
3. quite often the user wants to edit things in some temporary space and "commit" later via some yes/no dialog, which would mean creating another relatively complicated object with the need to move things with more code than a simple assignment.
4. a `select` option is a string so any code using it will need a list of string compares instead of a `switch()` everywhere.

I will think about it. The `number` is of course less controversial and a `switch` as a special case of the "boolean" `enum` probably also makes sense if the globals are abandoned.

I am also planning to add a lambda for delegating the contents of the displayed line to the user in case some advanced formatting is needed.

Thanks again

numo68 committed:
Change the edited variable to a number object

numo68 committed:
Change enum to use select component

numo68 committed:
Add lambda for the menu item

numo68 committed:
Minor fixes

numo68 committed:
Another formatting fix

numo68 commented:
@jesserockz I implemented the `number` and `select` and a lambda for user-defined formatting. I did not do `switch` as I think there is no possibility to name the on and off values (I don't want to use fixed English names). Do you think it makes sense to implement a `switch` toggle with adding the strings for on and off as a config, or should the `switch` itself maybe be extended? Thanks.

jesserockz commented:
>  I implemented the number and select and a lambda for user-defined formatting

Awesome, Ill take another look.

> Do you think it makes sense to implement a switch toggle with adding the strings for on and off as a config

Yes, config options for `on_text` and `off_text` make sense here.

numo68 committed:
Implement switch

numo68 committed:
fix formatting

numo68 commented:
``switch`` is now implemented as well.

jesserockz commented on esphome/components/lcd_menu/__init__.py:
```diff
@@ -0,0 +1,338 @@
+                f"{CONF_SWITCH} has to be present if {CONF_TYPE} is '{CONF_SWITCH}'"
+            )
+
+    return config
```
Most of this can be replaced by using `cv.typed_schema`

numo68 replied:
oh, did not know this one. Changed to `typed_schema`

nagyrobi commented:
How lovely would be to have this possible on Oled / Epaper displays too. 
Like a placeable object. Inside a rectangle.

mmakaay commented:
> 2\. from the bird's eye view the `select` is an _unordered_ set, there is no concept of cycling through options built in. One would need to find it in the list (by the way, `SelectTraits::get_options()` is not returning a reference) and work from there.

I submitted a PR for making the select component more useful in that way. Guess what? It's for making it simple to control a select component through for example a rotary encoder ;-) 
See: https://github.com/esphome/esphome/pull/3423

numo68 commented:
> How lovely would be to have this possible on Oled / Epaper displays too. Like a placeable object. Inside a rectangle.

I know and I have a use case for pixel-oriented displays as well, good to see that I am not alone. :) The code is written with that somehow in mind; the methods actually writing things are not that many. Unfortunately there is no common class representing both so at the moment this would need another display object and changed behavior according to which one is set.

If/when this is merged, I might work on refactoring it so that the common parts interacting with the variables and the basic navigation logic get isolated and create separate objects for driving character- and pixel-based displays, adding another integration for the latter. This was too much to consider as a single PR, for example things are not that easy anymore when variable-width fonts are taken into account, people will want to add borders, draw the selected line inverted, defining colors etc so this needs a proper design. E-paper might also have challenges regarding the update speed as well (I don't have any to experiment with).

numo68 committed:
use cv.typed_schema

numo68 commented:
As the possible extension for multiple display types involves configuration change I would like to do the basic structure as part of this PR to avoid compatibility break later. I guess I should aim for something like

```
display_menu:
  - platform: lcd_character
    display_id: foo
    menu: ...
  - platform: ...
    ...
```

What is the canonical example of such structure with a common idea and multiple implementations where I can borrow the correct idioms like how to do namespaces (one for all or one for base and one per platform?), how to combine the schemas and code generation of base and platforms etc? My gut feeling tells me that `output` with a subdirectory per platform is intended for this kind of stuff (rather than e.g. the `display` involving `display.py` and specializations on the same directory level as the base) but I'd like to verify first. Thanks.

numo68 committed:
refactor lcd_menu to common and lcd char-based part

numo68 committed:
fix typos and style

nagyrobi commented:
One question: in the docs, you say _the I²C one running at the speed according to the datasheet will create perceptible delays especially when changing a numeric value using the rotary encoder. Most PCF8574 adapters used with these displays will happily run at 200 or even 400 kHz though so if you are comfortable accepting risks from running your hardware out of spec, you might want to try that._

I don't see any option to change the frequency in the PCF config, how do you suggest experimenting with that?

numo68 commented:
Alright, I tried to rearrange things a bit. There is now a `display_menu_base` with the common part and the `lcd_menu` integration implementing the parts specific for character based displays. The configuration stays the same, with just the automations now being called `display_menu.*` instead of `lcd_menu.*` (I did not update the documentation PR yet).

By the way, is there a way to run _all_ the github checks locally? The `docker ... esphome/esphome-lint ` does not seem to run everything no matter what script I try running, quicklint is complaining of things that are there (not finding platform includes), on the other hand it is able to mess things up (changes names of virtual methods in the base but not in the derived class etc). I guess some of this might be due to attempting to use docker after compiling locally so if there is something cached, the paths won't match, but a documented way including common pitfalls would be really helpful here.

Thanks

numo68 committed:
fix style

numo68 commented:
> I don't see any option to change the frequency in the PCF config, how do you suggest experimenting with that?

This needs to be applied for the whole bus and set to the speed of the slowest device there as every device needs to at least understand that the message is not meant for it. My config now contains

```
i2c:
  sda: GPIO21
  scl: GPIO22
  frequency: 200kHz
```

For me the 400 kHz (usual highest speed on I2C) was completely fine as well, with the stock adapter bought somewhere on amazon and 20 cm long jump wires going through a 3.3V to 5V level shifter so not exactly a best environment ever :) The datasheets of the common manufacturers tend to specify 100 kHz (common "low-speed" I2C).

nagyrobi commented:
Ah, OK. You might want to make a small clarification for the masses: _...you might want to try that in your `i2c` config._

jesserockz commented on esphome/components/display_menu_base/display_menu_base.h:
```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include "esphome/components/number/number.h"
+#include "esphome/components/select/select.h"
+#include "esphome/components/switch/switch.h"
```
Any code touching `number` `select` or `switch` needs to be wrapped with `#ifdef USE_NUMBER` etc.
For example, if someone has the menu in their yaml, but no `number` components, then the `number` header will not exist in the generated project and will fail to compile.

numo68 replied:
done

jesserockz commented on esphome/components/display_menu_base/display_menu_base.h:
```diff
@@ -0,0 +1,150 @@
+#include "esphome/components/number/number.h"
+#include "esphome/components/select/select.h"
+#include "esphome/components/switch/switch.h"
+#include "esphome/components/lcd_base/lcd_display.h"
```
Similar to the above, this `display_menu_base` should not have a dependency on things like `lcd_menu`. It should define an interface that the `lcd_menu` would implement and use that to call any virtual functions.

jesserockz replied:
Strike that, I read wrong here, but shouldn't any `lcd_display` code be in the `lcd_menu` code?

numo68 replied:
Correct, the include is not needed (I forgot to remove that while splitting to the common and specific part). Will fix that.

numo68 committed:
add USE_* ifdefs, cleaned up includes

numo68 committed:
fix style

nagyrobi commented:
Do I understand well that currently this implementation is limited explicitly to _select_, _number_ and _switch_ only?
How hard would it be to cover any kind of platform? In practice, users will want to adjust direct outputs, fan speeds, climate target temperatures, light dimmers or RGB levels etc. with floats, and tun on and off many other stuff not just switches.

I know this can be worked around with a multitude of template selects, numbers and switches acting with automations on the desired devices as proxies, but how about using lambdas to control them directly? Would that be possible? Some easy way to get/set the value with `x`?

numo68 commented:
> Do I understand well that currently this implementation is limited explicitly to _select_, _number_ and _switch_ only?

And a `command` just firing the `on_value`, yes. 

> How hard would it be to cover any kind of platform? In practice, users will want to adjust direct outputs, fan speeds, climate target temperatures, light dimmers or RGB levels etc. with floats, and tun on and off many other stuff not just switches.
> 
> I know this can be worked around with a multitude of template selects, numbers and switches acting with automations on the desired devices as proxies, but how about using lambdas to control them directly? Would that be possible? Some easy way to get/set the value with `x`?

Hm... I think the main problem is with the metadata, the number needs the min/max/step, the select needs the list of the options etc. My initial proposal was operating on the globals and had those metadata as configuration, the proposal to use `select`, `switch` and `number` was then made by Jesse and indeed looks nicer as those are generic enough to not pull additional logic in.

What I can however well imagine is a `custom` menu item abstracting the control interface, so that `x` would be completely in the control of the user defined code. You'd get `on_clockwise`, `on_counterclockwise`, `on_enter` and have to provide a lambda for getting the value to display. Would that cover the needs? @jesserockz what is your opinion?

nagyrobi commented:
For things like climate presets or modes, or light effects, which don't resemble natively in a select, would be nince to have a lambda-like way to set those. Even if not so automagic like with the native select (eg. you'd have to manually repeat the available climate modes in both climate config and the lcd menu config - that would be fine).

nagyrobi commented:
Custom with read and write lambdas should be OK.

numo68 committed:
move text value handling to the item

numo68 committed:
make text templatable

numo68 committed:
move next/prev logic to the menu item

numo68 committed:
add custom menu item

numo68 committed:
fix style

numo68 committed:
style fixes

numo68 commented:
Implemented in the last round of commits:
- `text` is now templatable
- there is a `value_lambda` allowing the user to define own value-to-string conversion for all editable item types (`select`, `number`, `switch`, `custom`)
- there is a `custom` type delegating changes of the value to `on_next` / `on_prev` automations

The related documentation PR esphome/esphome-docs#2029 was updated as well.

numo68 commented:
FWIW what I plan to do next is
- a joystick mode
  - back button
  - reverse logic as with the encoder clockwise means down in menu/increment value, the up button means up in menu/increment value
  - immediate edit mode for numbers as well (up/down navigates, left/right changes the value)
- maybe do first tests with pixel-oriented LCD to explore the terrain and pitfalls

numo68 committed:
implement joystick mode

numo68 committed:
Merge remote-tracking branch 'upstream/dev' into display/lcd-menu

numo68 committed:
use the new way of cycling options and inc/dec numbers

numo68 committed:
fix style

numo68 commented:
The joystick mode is implemented and the `select` and `number` now use the new way of iterating / in-/decrementing (perfect timing :)).

The pixel-oriented displays have more things to consider, especially if using fonts that are not monospaced and the goal is to have the menu confined to a rectangle instead of using the whole display. Something like that will use more time to get right and if I start working on that, I'll fork from this branch to not to spam this PR with experimental commits. Frankly I think that if the ESPHome should go in the direction of more advanced GUIs pretty much all the `DisplayBuffer`'s drawing methods should get a version including a bounding box to start with and that would be the first PR to implement. How to do that without growing the interface to an impractical size and without consuming the CPU cycles at each pixel drawn in the 99% it is not needed is not a trivial task. Who is the best person to ask here? - there is quite a lot of people committing there and no entry in CODEOWNERS.

@jesserockz that means that I am finished with the functional scope of this PR and I would appreciate reviewing it in order to merge when it is ready for it. The documentation is updated as well. Thanks.

jesserockz commented:
Please don't add more to this already large PR :smile:. I will review it once the stable release goes out next week.

nagyrobi commented:
> Frankly I think that if the ESPHome should go in the direction of more advanced GUIs pretty much all the `DisplayBuffer`'s drawing methods should get a version including a bounding box to start with and that would be the first PR to implement. How to do that without growing the interface to an impractical size and without consuming the CPU cycles at each pixel drawn in the 99% it is not needed is not a trivial task.

Should check the OpenHASP project...

nagyrobi commented:
> The joystick mode is implemented

This should work better with i2c, right? Not too prone to so fast changes than with rotary?
Think of using a 5-way button pad also with i2c to save gpios on the mcu.

numo68 commented:
> > The joystick mode is implemented
> 
> This should work better than rotary with i2c, right? Not too prone to so fast changes than with rotary?

That's probably right, definitely with knobs with a "paddle" allowing to fully rotate the encoder using a single finger. With a plain knob you are forced to lift your fingers after a few events (seems to be around 9 on mine) so I don't expect a massive difference compared to a good microswitch but there will be some.

jesserockz commented on esphome/components/display_menu_base/display_menu_base.cpp:
```diff
@@ -0,0 +1,554 @@
+
+#ifdef USE_NUMBER
+bool MenuItem::inc_number_() {
+  bool chg = false;
```
I just chose a random one, but please use full names for all variables, it only serves as confusion when trying to read the code.
```suggestion
  bool changed = false;
```

jesserockz commented on esphome/components/display_menu_base/__init__.py:
```diff
@@ -0,0 +1,398 @@
+        default_type="label",
+        lower=True,
+    ),
+    validate_menu_item,
```
Since this only contains a validation for number format, it could be directly on the schema validation for CONF_FORMAT

jesserockz commented on esphome/components/display_menu_base/automation.cpp:
```diff
@@ -0,0 +1,10 @@
+static const char *const TAG = "display_menu.automation";
+
+}  // namespace display_menu_base
+}  // namespace esphome
```
No need for an empty file.

jesserockz commented on esphome/components/display_menu_base/display_menu_base.cpp:
```diff
@@ -0,0 +1,554 @@
+static const char *const TAG = "display_menu";
+
+void DisplayMenuComponent::setup() {}
+
```
No need to override if its not even used
```suggestion
```

jesserockz commented on esphome/components/display_menu_base/display_menu_base.cpp:
```diff
@@ -0,0 +1,554 @@
+#include "display_menu_base.h"
+#include "esphome/core/hal.h"
```
What is this imported for in this file?

jesserockz commented on esphome/components/display_menu_base/display_menu_base.h:
```diff
@@ -0,0 +1,199 @@
+  void set_switch_variable(switch_::Switch *var) { this->switch_var_ = var; }
+  void set_on_text(const std::string &t) { this->switch_on_text_ = t; }
+  void set_off_text(const std::string &t) { this->switch_off_text_ = t; }
+#endif
```
It would be nicer if `MenuItem` was extended by subclasses for each type so that every MenuItem doesnt have every function and member.

numo68 replied:
done

jesserockz commented on esphome/components/display_menu_base/display_menu_base.h:
```diff
@@ -0,0 +1,199 @@
+  void set_mode(MenuMode mode) { this->mode_ = mode; }
+  void set_rows(uint8_t rows) { this->rows_ = rows; }
+
+  void setup() override;
```
Not needed
```suggestion
```

numo68 committed:
Merge branch 'dev' into display/lcd-menu

jesserockz commented on esphome/components/lcd_menu/__init__.py:
```diff
@@ -0,0 +1,72 @@
+        {
+            cv.GenerateID(): cv.declare_id(LCDCharacterMenuComponent),
+            cv.GenerateID(CONF_DISPLAY_ID): cv.use_id(lcd_base.LCDDisplay),
+            cv.Required(CONF_DIMENSIONS): validate_lcd_dimensions,
```
This can be inherited from the display. See uses of `inherit_property_from`

jesserockz reviewed (changes requested):
Generally looks good, just a few things below.

numo68 committed:
use explanatory names, remove unneeded things

numo68 committed:
simplify menu item validation

numo68 committed:
inherit the dimensions from the parent display

numo68 commented:
Addressed all comments except the `MenuItem` polymorphism as this needs some thoughts to select what is a good candidate for a virtual method and what is not. The advantage is indeed cleaner code and a bit less memory used, the disadvantage is that things like `get_number_value()` need a cast as the automations just get a `MenuItem *`, the `__init__.py` is more complicated due to allocating different classes and lacking rtti one probably cannot get rid of the `MenuItemType` altogether as the `dynamic_cast` is unavailable (but let's see). I'll take a look probably next week. Thanks.

numo68 committed:
split menu component and menu item files

numo68 committed:
refactor menu item to polymorphic classes

nielsnl68 commented:
hi @[numo68](https://github.com/numo68) just wondering did you test the `text' option with a !lambda result?
I made my own component that uses yours as inspiration and when i tested the compiler did not liked it at all.

numo68 commented:
> hi @[numo68](https://github.com/numo68) just wondering did you test the `text' option with a !lambda result?
> 
> I made my own component that uses yours as inspiration and when i tested the compiler did not liked it at all.


I did, it should even be part of the test1.yaml that gets built automatically at commit. I am however not at the computer so I cannot double-check.

nielsnl68 commented:
> 

Oww yes, i missed that yesterday. Sadly on my attempt with my component/hub it does not like it.
Maybe i have missed something important. when you have time, when you behind your PC again. could you check on discord and my thread about this issue: https://discord.com/channels/429907082951524364/980094366519541812/980095246140588042
thanks.

numo68 committed:
fix style

numo68 commented:
finished the `MenuItem` polymorphism as well.

nagyrobi commented:
One observation, happened to me while testing. Please handle gracefully the case when the display is missing at boot.

Currently if the display is missing (say, I2C bus cable broken, pcf chip disconnected etc.) the node hangs completely when lcd_menu tries to access it.

This should be handled during boot, if display component reports no display, menu shouldn't either, but rest of the things should keep operational.

Reason: implementing this on a node controlling the entire heating system of the house. The heating controls, thermostats, automations etc. should still be usable (via other means like HA) even if the display goes out of order. This seems to be OK without the menu, but as soon as it's added to the config, it won't...

numo68 commented:
> This should be handled during boot, if display component reports no display, menu shouldn't load, but just print an error message to the log and rest of the things should keep operational.

Not sure that the lcd_menu knows that. I am just the user, no different that the menu's lambda. I'll take a look but I am afraid the robustness in case of hw failures is not necessarily the ESPHome's strenghth...

> This seems to be OK without the menu, but as soon as it's added to the config, it won't...

Do you mean that a mere print does not hang the system but a draw of the menu does?

nagyrobi commented:
When I touch the rotary, it hangs...

numo68 commented:
> When I touch the rotary, it hangs...

Could you try not using the `lcd_menu` and do just a `it.print("Hello World!");` from the display's lambda? It is also possible that this is in fact a timing issue - a change of the display's content does an immediate `update`, instead of the automatic periodic one.

In any case, I have looked and did not found any method exporting the health status of a `PCF8574LCDDisplay` nor `LCDBase` object. Handling a broken display gracefully has to go into `LCDDisplay::display()` - not sure who owns that, there is nobody in CODEOWNERS...

jesserockz commented:
You can check `{display}->is_failed()` in the `lcd_menu` setup and then fail the `lcd_menu` setup early too.
This only works though if the parent display actually sets a failure during its setup, which `lcd_pcf8574` does.

nagyrobi commented:
@jesserockz how about `lcd_gpio`?

numo68 commented:
> You can check `{display}->is_failed()` in the `lcd_menu` setup and then fail the `lcd_menu` setup early too. This only works though if the parent display actually sets a failure during its setup, which `lcd_pcf8574` does.

Thanks, I did not know about this, I will implement it. That won't help if the wire breaks later though.

As I need to ensure the menu setup running after the display one, is it a good style to derive own setup priority from that of another component or should it be a static number of the display (PROCESSOR) with an offset? Not many examples of either unfortunately.

An unrelated question: what is the position of the core developers on const-correctness, is it a goal to converge to, i.e. are such patches accepted? From my daily work I am used to be strict with `const` as it helps both the readability and the compiler but ESPHome often does not care (see e.g. `bool is_failed();`) and I have to either give up or be constantly (pun intended) bumping into a compiler error and needing an ugly `const_cast` as a result.

nagyrobi commented:
IMHO it's fine if at boot/setup it's detected if it's failed, as even if it crashes in the middle somewhere, it will eventually be rebooted and from there at least the rest of the functionality can be regained.

As it happened to me yesterday: the daughterboard on the LCD screen failed after a few minutes. It can't be detected anymore on the bus. Examined more close and discovered that the chip on it is not a PCF8574, but an AT8574, which seems like a [clone](http://www.analogtek.com.cn/products/show-212.html) of the original PCF chip. Unfortunately it ceased functioning after a few minutes, can't do anything to bring it back to life.

numo68 committed:
Disable the menu on the failure of the display

numo68 commented:
> IMHO it's fine if at boot/setup it's detected if it's failed

Implemented. The menu is disabled completely (i.e. not display only) to avoid the user blindly interacting with the system

> Would be nice if it would print an error log message which could be seen when the API logger connects.

```
17:59:57][C][lcd_pcf8574:024]: PCF8574 LCD Display:
[17:59:57][C][lcd_pcf8574:025]:   Columns: 20, Rows: 4
[17:59:57][C][lcd_pcf8574:026]:   Address: 0x27
[17:59:57][C][lcd_pcf8574:027]:   Update Interval: 1.0s
[17:59:57][E][lcd_pcf8574:029]: Communication with LCD Display failed!
[17:59:57][C][lcd_menu:022]: LCD Menu
[17:59:58][C][lcd_menu:023]:   Columns: 20, Rows: 4
[17:59:58][C][lcd_menu:025]:   Mark characters: 3e, 2a, 7e, 5e
[17:59:58][E][lcd_menu:027]: The connected display failed, the menu is disabled!```
```

> As it happened to me yesterday: the daughterboard on the LCD screen failed after a few minutes. It can't be detected anymore on the bus.

Oh. Probably stating the obvious but by the way, don't forget that most of these LCDs are 5V while the ESP32 uses 3.3V so a level converter is needed. A forgotten one can do anything from destroying things to non-working comms.

nagyrobi commented:
> Implemented. The menu is disabled completely (i.e. not display only) to avoid the user blindly interacting with the system

Perfect! Very proffessional!


> most of these LCDs are 5V while the ESP32 uses 3.3V so a level converter is needed.

Really? I thought power is at 5V, I2C pins shouldn't stay at 3.3V level??

nagyrobi commented:
Yeah looking at the datasheet (of the genuine PCF8574) it should work at 3.3V if supply voltage is also 3.3V, so I tried that, and this AT8574 is still not detected at all.
I've attached to the same pins other I2C devices (including a PCF8574 on a GPIO-expander module), they are all detected by the ESP, only the clone isn't.

nagyrobi commented:
Removed the faulty expander and hooked the display directly via GPIO.

I see one issue, when rendering texts from lambda. For example this should render as `°C`, but instead, it shows a weird character:
```
    - type: label
      text: !lambda |-
        return "\xdfC
```
If I add a space, like `return "\xdf C`, it wil render with space, like `° C`.
So it looks like hex strings are mangled with next characters and decoded errorneously.
I have to define all the `user_characters` because my language has lots of diacritics, it's impossible to properly print texts because of this.

nagyrobi commented:
Question:
How to display a sensor value in the label? Like:
`        return printf("F\x03t\x05s: %8.0f\xdf""C", id(number_test).state);`
throws:
```
/config/esphome/wt32-eth01_LCD_rotary.yaml: In lambda function:
/config/esphome/wt32-eth01_LCD_rotary.yaml:219:20: error: could not convert 'printf(((const char*)"F\003t\005s: %8.0f\37777777737C"), ((double)number_test->esphome::template_::TemplateNumber::<anonymous>.esphome::number::Number::state))' from 'int' to 'std::__cxx11::string {aka std::__cxx11::basic_string<char>}'
         return printf("F\x03t\x05s: %8.0f\xdf""C", id(number_test).state);
                    ^
*** [/data/wt32-eth01-lcd/.pioenvs/wt32-eth01-lcd/src/main.cpp.o] Error 1
```

(taken from [doc of LCD](https://deploy-preview-2029--esphome.netlify.app/components/display/lcd_display.html#rendering-lambda), goal was _right-padded sensor value with 0 digits after the decimal_)

numo68 commented:
> my language has lots of diacritics

Oh yes, mine as well ;) Unfortunately no luck with these LCD displays...

> EDIT: looks like `return "\xdf""C";` will do the trick...

Exactly. C is a hex number, and the C standard does not define the maximum number of them in an `\x` escape sequence, so the C is interpreted as part of it.

> return printf("F\x03t\x05s: %8.0f\xdf""C", id(sensor).state);

printf (a different thing from it.printf of the LCD) does not return a string. Use snprintf into a char[] and return std::string(that). Unfortunately there is no easy to use C++ formatting library until the C++-20 version of the standard which I think the toolchain used for ESPhome does not support yet, at least it did not when I tried.

nagyrobi commented:
> Unfortunately no luck with these LCD displays

As far as I see, these displays seem to be driven by [HD44780 Hitachi character generator](https://en.wikipedia.org/wiki/Hitachi_HD44780_LCD_controller), and there should exist two versions of it:

- HD44780UA00, the standard (Japanese) version, which includes katakana characters and some Greek letters and mathematical - HD44780UA02, a European version, which includes the complete set of Greek, Cyrillic and Western European characters (with diacritics)

Seems that Aliexpress stuff is based on HD44780UA00, couldn't find a listing explicitly stating which version of controller is on board. HD44780UA02 would be much better for me:
![image](https://user-images.githubusercontent.com/1550668/173113487-9c98e866-8ee4-4a3c-a83f-61fe62057c5f.png)

nagyrobi commented:
A suggestion for a better `mark_back` character: ![image](https://user-images.githubusercontent.com/1550668/173116313-81e390fd-8716-4f92-b300-be16da413d28.png)

```
        - 0b00100
        - 0b01000
        - 0b11110
        - 0b01001
        - 0b00101
        - 0b00001
        - 0b11110
        - 0b00000
```

nagyrobi commented:
> Use snprintf into a char[] and return std::string(that).

Can you please show me an example? Doesn't have to be snprintf. My intention is to have a submenu with many sensor values shown, something like `Bedroom: 20°C` as labels.

numo68 commented:
> > Use snprintf into a char[] and return std::string(that).
> 
> Can you please show me an example? Doesn't have to be snprintf. My intention is to have a submenu with many sensor values shown, something like `Bedroom: 20°C` as labels. So how to embed a sensor value in a lambda.

```
    - type: label
      text: !lambda |-
        char buf[16];
        snprintf(buf, sizeof(buf), "Bedroom: %4.1f\xdf""C", id(my_number_1).state);
        return std::string(buf);
```

> Questions:
> 
> * would it be possible to use the `user_characters` not only in lambdas? Simply in `text` option...

Sorry, I don't know how to get the `\x01` and similar to the yaml configuration. I guess it is possible somehow. Maybe someone else will step in.

> * I see examples for `switch`, etc. how about triggering a `button` press?

Use `command`'s `on_value` (not sure if simulating a button press is possible though but you can execute actions). See the linked esphome-docs PR (generated at https://deploy-preview-2029--esphome.netlify.app/components/display_menu/index.html ) for the complete documentation; the config here is just an example.

nagyrobi commented:
> See the linked esphome-docs

Thanks. Testing each one. Found an error in the `switch` doc, it reads `select: my_switch`.

nagyrobi commented:
My screen has dimensions 16x2. It's configured as `dimensions: 16x2`.
But the menu renders menu items at 15x2. Last character position is not used, except when when there are submenus, then it shows an `mark_submenu`, or the _Menu is not active_ message is shown.
Would be nice if the line's last character wouldn't be lost when there aren't submenus, as those are usually more.

numo68 commented:
> Would be nice if the line's last character wouldn't be lost when there aren't submenus, as those are usually more.

That is by design and in line with what I have seen in other similar implementations (such as my 3D printer). The values' right brackets interspersed with submenu marks simply look ugly to me. So unless more people wish for that so an option for that is warranted, that will stay.

nagyrobi commented:
What do you mean by 
> The values' right brackets interspersed with submenu marks
?

numo68 commented:
> What do you mean by
> 
> > The values' right brackets interspersed with submenu marks
> 
> ?

To me the

```
 Back           ^
 Submenu 1      >
 Quite Big Label
 Setpoint [15.0]
 Submenu 2      >
```

looks better than

```
 Back           ^
 Submenu 1      >
 Some Large Label
 Setpoint  [15.0]
 Submenu 2      >
```

nagyrobi commented:
My use case: I have submenus listing all the current values measured, I really have to struggle to fit the lines in. For example temperatures (similar for humidities, sensors battery levels etc):
```
>Bedroom:  16˚C
 Livingroo:16˚C
 Bathroom: 16˚C
 Outdoors: 16˚C
 Kitchen:  16˚C
 Petersroo:16˚C
 Annasroom:16˚C
 Back          ^
```
This looks ugly too, only the `Back` option (which has the shortest text btw) is using the last character, whilst I could use that extra missing one in the labels.

numo68 commented:
> Would be nice to be a configurable option so one could set this per needs (eventually change it per menu?) so one could use both ways?

Understood. I'll wait for Jesse's review as there was a request not add more things in and there were even doubts regarding the templated text in a discord chat related to the Niels's issue a few days ago (seems like that I was the first to use a templated string for the configuration item like `text:` and not for an action ;) ) so let's see whether this is the only thing that can see some change. Seeing your use case maybe even a read-only `number`, `switch` and `select` would make sense, i.e. without the brackets and non-editable.

Anyway, those LCDs are indeed small and will always need compromises one way or another.

nagyrobi commented:
Must say I'm impressed by this feature and I will use it for at least two of my control systems (ventilation control and heating control). I will be putting thermostats, fans, numbers, switches on local LCD controls, and display statuses of lots of sensors.

You can count on me for ideas and testing.

nagyrobi commented:
For some of my new ideas, what would be more appropriate?
- pour them inhere, or create normal feature requests?
- put them all in a single feature request as a list, or should I create separate for each?

nagyrobi commented:
How to boot with menu in inactive state?

The goal is: normally the node would boot with the display just showing statuses, a few sensors etc. With rotary long-press the user would enter the menu, view more details, change settings, etc. and with long-press again he would hide the menu back and the status screen will be shown.
I've implemented all this, however, I can't make the node to start with the menu disabled.
Adding to `esphome:`
```yaml
  on_boot:
  - priority: -650
    then:
      - display_menu.hide:
```
Will display the menu for a few seconds, and only disable after that.

numo68 commented:
> How to boot with menu in inactive state?

```
lcd_menu:
  ...
  active: false
```

> For some of my new ideas, what would be more appropriate?

Good question. I think that a discussion of the "Ideas" type is the right way - https://github.com/esphome/esphome/discussions/categories/ideas . I don't think there is much activity there but is probably better suited to things that are not part of an active PR (yet).

nagyrobi commented:
> a discussion of the "Ideas" type is the right way

I've opened a topic: https://github.com/esphome/esphome/discussions/3547

nagyrobi commented:
Putting this one here as later may be breaking change if considered.

How about renaming the `menu:` configuration parameter to something like `items:` to improve readability. I find too many times repeating the term `menu` close to each other and in different contexts, which creates a bit of confusion for the newcomer:

Currently:
```yaml
lcd_menu:
  id: my_lcd_menu
  ...
  menu:
    - type: menu
      text: !lambda return "Vez\x05rl\x09";
      menu:
        - type: switch
          immediate_edit: true
          text: !lambda return "Teker\x09";
          on_text: 'Bal'
          off_text: 'Jobb'
          switch: rotary_direction
        - type: command
          text: !lambda return "Ujraind\x07t\x06s";
          on_value:
            then:
              - button.press: reboot_normal
        - type: menu
          text: 'Safe Mode'
          menu:
            - type: menu
              text:  'Biztos benne?'
              menu:
                - type: command
                  text: 'Igen'
                  on_value:
                    then:
                      - text_sensor.template.publish:
                          id: text_lcd
                          state: "Safe Mode-ban"
                      - binary_sensor.template.publish:
                          id: print_message
                          state: ON
                      - display_menu.hide:
                      - button.press: reboot_safe
                - type: command
                  text: 'Nem'
                  on_value:
                    then:
                      - display_menu.show_main:
        - type: back
          text: 'Vissza'
```

How about:
```yaml
lcd_menu:
  id: my_lcd_menu
  ...
  items:
    - type: menu
      text: !lambda return "Vez\x05rl\x09";
      items:
        - type: switch
          immediate_edit: true
          text: !lambda return "Teker\x09";
          on_text: 'Bal'
          off_text: 'Jobb'
          switch: rotary_direction
        - type: command
          text: !lambda return "Ujraind\x07t\x06s";
          on_value:
            then:
              - button.press: reboot_normal
        - type: menu
          text: 'Safe Mode'
          items:
            - type: menu
              text:  'Biztos benne?'
              items:
                - type: command
                  text: 'Igen'
                  on_value:
                    then:
                      - text_sensor.template.publish:
                          id: text_lcd
                          state: "Safe Mode-ban"
                      - binary_sensor.template.publish:
                          id: print_message
                          state: ON
                      - display_menu.hide:
                      - button.press: reboot_safe
                - type: command
                  text: 'Nem'
                  on_value:
                    then:
                      - display_menu.show_main:
        - type: back
          text: 'Vissza'
```

numo68 committed:
rename menu: to items:

numo68 committed:
Merge remote-tracking branch 'upstream/dev' into display/lcd-menu

numo68 commented:
> How about renaming the `menu:` configuration parameter to something like `items:` to improve readability. I find too many times repeating the term `menu` close to each other and in different contexts, which creates a bit of confusion for the newcomer:

Makes sense, done. Documentation PR is changed as well.

nagyrobi commented:
don't lest this stale please :)

numo68 committed:
Merge remote-tracking branch 'upstream/dev' into display/lcd-menu

jesserockz committed:
Merge branch 'dev' into display/lcd-menu

jesserockz committed:
Update __init__.py
